jupyterhub/tests/test_scopes.py

```diff
@@ -648,3 +648,62 @@ async def test_server_state_access(
             assert r.status_code == 403
         app.db.delete(group)
         app.db.commit()
+
+
+@mark.parametrize(
+    "name, user_scope, token_scope, discarded_own_scope, retained_owner_scope",
+    [
+        ('no_filter', 'users:activity', 'users:activity', False, False),
```

Comment:
More combinations to test:

- subscopes cross-filter users!user=x, read:users:name
- multiple user filters ([users!user=x, users!user=y] vs [read:users:name!user=y])
- users!group=y vs users!user=x (won't do what we *want*, but good to have a test to ensure it does what we *expect*)
- users *and* groups, with user match: `users!group=x, users!user=y` vs `read:users!user=y, read:users:name!group=x`

Reply:
Thank you, the combinations were added.  

There is one more with `!user=` and `!server=` treated in the same way as `!group=` and `!user=` (no intersection with warning). As now any scope containing the `!server=` filter gets 'expanded' with `read:users:name!user=` scope, I think the `read:users:servers` should then have `read:users:name` as a subscope. It has been added in this PR to make sure that a user with `users:servers!user=x` scope is allowed to have token with `users:servers!server=y` (which checks for all subscopes ignoring horizontal filters) but their intersection (for now) returns an empty set.

The scope hierarchy needs to be updated according to latest docs (e.g., `admin:users` scopes should contain `users` scope), this will be added in tiny separate PR.